### DISCUSSION TITLE
is there a way to use the ~/.ssh config file with multiple source control repos (GitHub, google Cloud Source,etc)

### DISCUSSION THREAD

**Discussion by dmastrop** [community/community#60727] [Programming Help] [answered]
### Body

hi All

I have a question on authenticating with multiple source control repos 
In my case I have a GitHub repository that I am working with and I need to add Google Cloud Source repository so that I can use my PC with both (using different terminal sessions of course)

For  kubectl there is a KUBECONFIG env variable that can be set per terminal that allows something similar to be done so that terminals can be used with different clouds, docker desktop, etc.
For example with default terminal set to ~/.kube/config one can set a new terminal to

export KUBECONFIG=~/my_new_config_for_this_terminal/config and kubeconfig will work with that new config for that terminal session.

For Google Cloud Source there is a private ssh key that needs to be used to authenticate to the Cloud Source repo.

The default terminal has ~/.ssh/config set with GitHub repo stuff (see below) and I need that for the default terminals.

Is there a SSHCONFIG path env variable or something similar that can be used so that I can set the terminal that is using Google Cloud Source repo to use that different .ssh config file?

The current .ssh/config file is this:

Host github.com
Hostname ssh.github.com
Port 443
User git

the .ssh/config for Google cloud source authentiation looks like this: (wildcard host, as we don't know what the hostname will be). (id_gcp_cloud_source is the private ssh key that was generated for this earlier)

Host *
  AddKeysToAgent yes
  UseKeychain yes
  IdentityFile ~/.ssh/id_gcp_cloud_source

If I just append the google config .ssh file to the current .ssh/config file the GitHub terminal will match both and I don't want that (second Host stuff will be added to git first entry).  I don't know if this will work or not, but I don't want it to break the GitHub repo terminal sessions.

I know there are probably a lot of folks doing something similar. What is the process for this?

warmest regards

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by nirav-gajera** [comment, marked as answer]
- To authenticate with multiple source control repositories, such as GitHub and Google Cloud Source, you can utilize SSH configuration and different SSH keys for each repository.
- Create separate SSH key pairs: Generate a new SSH key pair for each repository you want to authenticate with. For example, you can generate one key pair for GitHub and another for Google Cloud Source. You can use the ssh-keygen command to create the keys.
- Update your SSH configuration: Modify your ~/.ssh/config file to include the configurations for each repository.
 
-ex:  - GitHub
Host github.com
  Hostname ssh.github.com
  Port 443
  User git
  IdentityFile ~/.ssh/github_key

-Google Cloud Source
Host gcs
  Hostname gcs.example.com
  User your_username
  IdentityFile ~/.ssh/gcs_key

- Use SSH config per terminal: When using different terminals, you can set the SSH_CONFIG environment variable to specify the path to a different SSH configuration file. For example: `export SSH_CONFIG=~/path/to/google_ssh_config
`

-By setting the SSH_CONFIG environment variable, you can use a different SSH configuration file for that specific terminal session. This allows you to have separate configurations for each repository while avoiding conflicts between them.

**Reply by dmastrop** [reply]
thanks Nirav
warm regards

**Reply by dmastrop** [reply]
Sure. Thanks again for the thorough answer. warm regards
